cyipopt/tests/unit/test_ipopt_funcs.py

```diff
@@ -191,6 +191,87 @@ def intermediate(
     np.testing.assert_allclose(x_iterates[-1], x)
 
 
+@pytest.mark.skipif(
+    pre_3_14_0,
+    reason="GetIpoptCurrentIterate was introduced in Ipopt v3.14.0",
+)
+def test_get_iterate_hs071_12arg_callback(
+    hs071_initial_guess_fixture,
+    hs071_definition_instance_fixture,
+    hs071_variable_lower_bounds_fixture,
+    hs071_variable_upper_bounds_fixture,
+    hs071_constraint_lower_bounds_fixture,
+    hs071_constraint_upper_bounds_fixture,
+):
+    x0 = hs071_initial_guess_fixture
+    lb = hs071_variable_lower_bounds_fixture
+    ub = hs071_variable_upper_bounds_fixture
+    cl = hs071_constraint_lower_bounds_fixture
+    cu = hs071_constraint_upper_bounds_fixture
+    n = len(x0)
+    m = len(cl)
+
+    #
+    # Define a callback that uses some "global" information to call
+    # get_current_iterate and store the result
+    #
+    x_iterates = []
+    iter_counts = []
+    def intermediate(
+        alg_mod,
+        iter_count,
+        obj_value,
+        inf_pr,
+        inf_du,
+        mu,
+        d_norm,
+        regularization_size,
+        alpha_du,
+        alpha_pr,
+        ls_trials,
+        problem,
```

Comment:
Instead of counting arguments, can this last argument be a kwarg? Then if `problem == None` you get old behavior?

Reply:
The user can implement a callback function with `problem=None` as the last argument. To support this, we would need to send `problem=self` to the user's callback, which would still break existing callbacks. So in this case we would still need to branch on the signature of the provided callback.

We could update the "new call signature" to expect `**kwds`, which would make it more extensible to future arguments, at the expense of needing to document the keywords that we provide.

Reply:
In general, I don't find the original design pattern of 1) create a class with some methods and 2) pass this class to Problem a good idea. Requiring users to either subclass Problem or simply write normal functions that evaluate the various necessary elements cleaner approaches. Seems like you're hacking around this bad design pattern. Would it be better to only support this new feature if you subclass Problem? Then we don't need a host of hacks.

Reply:
I agree that both of those are cleaner design patterns, but I still think expanding the call signature is valuable as it lets a user implement the callback in an easier and much less error-prone manner.

I'm thinking of a situation (the Pyomo CyIpopt wrapper) where the "problem object" (whether or not a subclass of Problem) is created programatically and the user provides a pure function for the callback. If the call signature is not expanded, the function needs to know about some "global" state to access the new methods.

Reply:
I guess there is no reason that a wrapper (sub)class can't just expect an augmented call signature from the user, then pass `self` to the user's callback so they can use the new methods.